File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;

int cnt = 0;

void dfs(vector<int> &numbers, int target, int idx, int sum)
{
    if (idx == (int) numbers.size())
    {
        if (target == sum)
            ++cnt;
        return ;
    }
    
    dfs(numbers, target, idx + 1, sum + numbers[idx]);
    dfs(numbers, target, idx + 1, sum + (numbers[idx] * -1));
}

int solution(vector<int> numbers, int target)
{
    dfs(numbers, target, 0, 0);

    return cnt;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    int _target;
    int[] numbs;
    int cnt = 0;
    
    void dfs(int idx, int val) {
        if (idx == numbs.length) {
            if (_target == val) {
                ++cnt;
            }
            
            return;
        }
        
        dfs(idx + 1, val + numbs[idx]);
        dfs(idx + 1, val - numbs[idx]);
    }
    
    public int solution(int[] numbers, int target) {
        _target = target;
        numbs = numbers;
        
        dfs(0, 0);
        
        return cnt;
    }
}